File: RadixCyclicNN/radixnet/vision.py
"""Images as text: a Stable Diffusion image encoder (the inverse of image generation) plus base64.

Stable Diffusion *generates* an image by decoding a latent - ``4 x H/8 x W/8``
numbers - with its VAE.  This module runs that process backwards: the same
VAE's **encoder** turns an image into its compressed latent (an 8x spatial
reduction into 4 channels: 48 times fewer numbers than the RGB pixels), the
latent is quantised to one signed byte per number and base64-encoded, and the
resulting text::

    img:sd:128x128:AAECAwQF...

is what the trigram network trains on and predicts.  :func:`decode_text` runs
the forward process again (base64 -> latent -> VAE decoder -> PNG), so an
encoded or *predicted* text can be looked at.

Backends
--------
* ``"sd"`` - the diffusion VAE (:class:`SDVaeEncoder`): needs ``pillow``,
  ``torch`` and ``diffusers`` plus the VAE weights (``$RADIXNET_SD_VAE``, a
  Hugging Face model id or a local directory; default
  ``stabilityai/sd-vae-ft-mse``, the fine-tuned Stable Diffusion VAE).
* ``"tiny"`` - :class:`TinyEncoder`: ``pillow`` only; the same 8x spatial
  reduction as an RGB thumbnail (3 channels), so the text format, the API,
  the CLI and the Images tab all work without the diffusion weights.
* ``"auto"`` (default) - ``sd`` when it can be loaded, else ``tiny``.

Both encoders are deterministic (the VAE's latent *mean* is used, not a
sample), so the same image always gives the same text.
"""

from __future__ import annotations

import base64
import io
import os
import re
import struct
import threading
from typing import Any

from .encoding import repair_base64

__all__ = [
    "DEFAULT_SD_VAE",
    "DEFAULT_SIZE",
    "ENCODERS",
    "LATENT_SCALE",
    "SDVaeEncoder",
    "TinyEncoder",
    "VisionError",
    "decode_text",
    "describe",
    "encode_image",
    "get_encoder",
    "pack_text",
    "parse_text",
]

HEADER = "img"
ENCODERS = ("auto", "sd", "tiny")
DEFAULT_SIZE = 128
"""Images are resized to ``size x size`` before encoding (a multiple of 8; 128 -> a 4 x 16 x 16 latent)."""
MIN_SIZE, MAX_SIZE = 8, 1024
LATENT_SCALE = 4.0
"""Quantisation range of the scaled SD latents: ``[-4, 4]`` maps to one signed byte."""
DEFAULT_SD_VAE = os.environ.get("RADIXNET_SD_VAE", "stabilityai/sd-vae-ft-mse")

_TEXT_RE = re.compile(r"^\s*img:([a-z0-9_]+):(\d+)x(\d+):(.*)$", re.S)


class VisionError(ValueError):
    """A missing dependency, an unreadable image, an unknown encoder or a text that is not an encoded image."""


# ---------------------------------------------------------------------------
# the text format
# ---------------------------------------------------------------------------


def pack_text(encoder: str, width: int, height: int, payload: bytes) -> str:
    """``img:<encoder>:<w>x<h>:<base64 of the quantised latent>``."""
    return f"{HEADER}:{encoder}:{width}x{height}:" + base64.b64encode(payload).decode("ascii")


def parse_text(text: str) -> tuple[str, int, int, bytes, bool]:
    """``(encoder, width, height, payload, repaired)`` of an encoded-image text.

    The base64 part is repaired when it is not clean - whitespace, stray
    characters, a missing tail (a predicted text may be cut off or garbled):
    junk is dropped, the padding is completed and ``repaired`` says so.  The
    payload is *not* padded here (the decoder pads or truncates it to the
    latent size it needs).
    """
    match = _TEXT_RE.match(str(text or ""))
    if not match:
        raise VisionError("not an encoded image: expected 'img:<encoder>:<w>x<h>:<base64>'")
    encoder, width, height, body = match.group(1), int(match.group(2)), int(match.group(3)), match.group(4)
    try:
        payload, repaired = repair_base64(body)
    except ValueError as exc:  # pragma: no cover - the junk filter makes this rare
        raise VisionError(str(exc)) from exc
    if width <= 0 or height <= 0:
        raise VisionError(f"invalid image size {width}x{height}")
    return encoder, width, height, payload, repaired


def _fit_payload(payload: bytes, length: int) -> tuple[bytes, bool]:
    """Pad with zeros or truncate to ``length``; returns ``(payload, changed)``."""
    if len(payload) == length:
        return payload, False
    if len(payload) > length:
        return payload[:length], True
    return payload + bytes(length - len(payload)), True


# ---------------------------------------------------------------------------
# images
# ---------------------------------------------------------------------------


def _pil():
    try:
        from PIL import Image
    except ImportError as exc:
        raise VisionError("image encoding needs Pillow: pip install pillow") from exc
    return Image


def load_image(data: bytes):
    """Decode image bytes (PNG, JPEG, WebP, ... whatever Pillow reads) into an RGB image."""
    Image = _pil()
    try:
        image = Image.open(io.BytesIO(data))
        image.load()
    except Exception as exc:  # noqa: BLE001 - Pillow raises many kinds
        raise VisionError(f"not a readable image: {exc}") from exc
    return image.convert("RGB")


def _check_size(size: int) -> int:
    size = int(size)
    if size < MIN_SIZE or size > MAX_SIZE or size % 8:
        raise VisionError(f"size must be a multiple of 8 between {MIN_SIZE} and {MAX_SIZE}, got {size}")
    return size


def _png_bytes(image) -> bytes:
    buf = io.BytesIO()
    image.save(buf, format="PNG")
    return buf.getvalue()


class TinyEncoder:
    """The stand-in: an RGB thumbnail at 1/8 of the size (3 channels, one byte each) - no diffusion weights needed."""

    name = "tiny"
    channels = 3

    def latent_shape(self, width: int, height: int) -> tuple[int, int, int]:
        return (self.channels, height // 8, width // 8)

    def encode(self, image, size: int) -> tuple[bytes, tuple[int, int, int]]:
        Image = _pil()
        small = image.resize((size // 8, size // 8), Image.LANCZOS)
        return small.tobytes(), self.latent_shape(size, size)

    def decode(self, payload: bytes, width: int, height: int):
        Image = _pil()
        shape = self.latent_shape(width, height)
        small = Image.frombytes("RGB", (shape[2], shape[1]), payload)
        return small.resize((width, height), Image.BICUBIC)


class SDVaeEncoder:
    """The Stable Diffusion VAE run backwards: image -> latent mean (``4 x H/8 x W/8``), quantised to bytes.

    ``model`` is a Hugging Face model id or a local directory holding a
    ``diffusers`` ``AutoencoderKL``; the weights load on first use (kept for
    the process).  ``device`` defaults to CUDA / MPS when torch sees one.
    """

    name = "sd"
    channels = 4

    def __init__(self, model: str | None = None, device: str | None = None) -> None:
        self.model = model or DEFAULT_SD_VAE
        self.device = device
        self.vae: Any = None
        self.scaling = 0.18215
        self.error: str | None = None
        self._lock = threading.Lock()

    @staticmethod
    def importable() -> bool:
        try:
            import diffusers  # noqa: F401
            import torch  # noqa: F401
        except ImportError:
            return False
        return True

    def latent_shape(self, width: int, height: int) -> tuple[int, int, int]:
        return (self.channels, height // 8, width // 8)

    def load(self) -> None:
        """Load the VAE (idempotent); raises :class:`VisionError` when the dependencies or the weights are missing."""
        if self.vae is not None:
            return
        with self._lock:
            if self.vae is not None:
                return
            try:
                import torch
                from diffusers import AutoencoderKL
            except ImportError as exc:
                self.error = "the Stable Diffusion encoder needs torch and diffusers: pip install torch diffusers"
                raise VisionError(self.error) from exc
            device = self.device
            if device is None:
                device = "cuda" if torch.cuda.is_available() else (
                    "mps" if getattr(torch.backends, "mps", None) and torch.backends.mps.is_available() else "cpu"
                )
            try:
                vae = AutoencoderKL.from_pretrained(self.model, torch_dtype=torch.float32)
            except Exception as exc:  # noqa: BLE001 - network, missing files, bad model ids
                self.error = f"cannot load the VAE {self.model!r}: {exc}"
                raise VisionError(self.error) from exc
            self.vae = vae.to(device).eval()
            self.device = device
            self.scaling = float(getattr(self.vae.config, "scaling_factor", 0.18215) or 0.18215)
            self.error = None

    def _to_tensor(self, image, size: int):
        import torch

        raw = torch.frombuffer(bytearray(image.tobytes()), dtype=torch.uint8)
        x = raw.to(torch.float32).view(size, size, 3).permute(2, 0, 1).unsqueeze(0)
        return (x / 127.5 - 1.0).to(self.device)

    def encode(self, image, size: int) -> tuple[bytes, tuple[int, int, int]]:
        import torch

        self.load()
        Image = _pil()
        image = image.resize((size, size), Image.LANCZOS)
        with torch.no_grad():
            posterior = self.vae.encode(self._to_tensor(image, size))
            latent = posterior.latent_dist.mean * self.scaling  # the deterministic inversion: no sampling noise
            q = torch.round(torch.clamp(latent / LATENT_SCALE, -1.0, 1.0) * 127.0).to(torch.int8)
        values = q.flatten().tolist()
        return struct.pack(f"{len(values)}b", *values), tuple(int(v) for v in latent.shape[1:])

    def decode(self, payload: bytes, width: int, height: int):
        import torch

        self.load()
        Image = _pil()
        shape = self.latent_shape(width, height)
        values = struct.unpack(f"{len(payload)}b", payload)
        z = torch.tensor(values, dtype=torch.float32, device=self.device).view(1, *shape) / 127.0 * LATENT_SCALE
        with torch.no_grad():
            sample = self.vae.decode(z / self.scaling).sample  # the forward image process
            pixels = ((sample.clamp(-1.0, 1.0) + 1.0) * 127.5).round().to(torch.uint8)
        data = bytes(pixels[0].permute(1, 2, 0).contiguous().flatten().tolist())
        return Image.frombytes("RGB", (int(pixels.shape[3]), int(pixels.shape[2])), data)


_ENCODERS: dict[str, Any] = {}
_ENCODERS_LOCK = threading.Lock()


def _instance(name: str) -> Any:
    with _ENCODERS_LOCK:
        enc = _ENCODERS.get(name)
        if enc is None:
            enc = _ENCODERS[name] = SDVaeEncoder() if name == "sd" else TinyEncoder()
        return enc


def get_encoder(name: str = "auto") -> Any:
    """The encoder ``name`` (``sd`` / ``tiny``), or with ``auto`` the diffusion VAE when it loads, else the stand-in."""
    key = (name or "auto").strip().lower()
    if key not in ENCODERS:
        raise VisionError(f"unknown image encoder {name!r}; expected one of: {', '.join(ENCODERS)}")
    if key == "tiny":
        _pil()
        return _instance("tiny")
    if key == "sd":
        enc = _instance("sd")
        enc.load()
        return enc
    _pil()
    if SDVaeEncoder.importable():
        enc = _instance("sd")
        if enc.vae is not None:
            return enc
        if enc.error is None:  # try once per process; a failed download is not retried on every image
            try:
                enc.load()
                return enc
            except VisionError:
                pass  # the weights are not available: the stand-in takes over (describe() reports the error)
    return _instance("tiny")


def describe() -> dict:
    """What is available: the dependencies, the configured VAE and which encoder ``auto`` resolves to."""
    def importable(module: str) -> bool:
        try:
            __import__(module)
        except ImportError:
            return False
        return True

    pillow = importable("PIL")
    sd = _ENCODERS.get("sd")
    torch_ok = importable("torch")
    diffusers_ok = importable("diffusers")
    return {
        "pillow": pillow,
        "torch": torch_ok,
        "diffusers": diffusers_ok,
        "sd_model": DEFAULT_SD_VAE if sd is None else sd.model,
        "sd_loaded": bool(sd is not None and sd.vae is not None),
        "sd_error": None if sd is None else sd.error,
        "encoders": list(ENCODERS),
        "default_size": DEFAULT_SIZE,
        "auto": None if not pillow else (
            "sd" if (sd is not None and sd.vae is not None) or (sd is None and torch_ok and diffusers_ok)
            or (sd is not None and sd.error is None and torch_ok and diffusers_ok) else "tiny"
        ),
        "text_format": f"{HEADER}:<encoder>:<w>x<h>:<base64 of one signed byte per latent number>",
    }


def encode_image(data: bytes, size: int = DEFAULT_SIZE, encoder: str = "auto") -> dict:
    """Image bytes -> ``{"text", "encoder", "width", "height", "latent_shape", "bytes", "chars", "source_size"}``."""
    size = _check_size(size)
    image = load_image(data)
    enc = get_encoder(encoder)
    payload, shape = enc.encode(image, size)
    text = pack_text(enc.name, size, size, payload)
    return {
        "text": text,
        "encoder": enc.name,
        "width": size,
        "height": size,
        "latent_shape": list(shape),
        "bytes": len(payload),
        "chars": len(text),
        "source_size": [image.width, image.height],
    }


def decode_text(text: str, encoder: str | None = None) -> dict:
    """Encoded (or predicted) text -> ``{"png", "encoder", "width", "height", "bytes", "repaired"}``.

    ``encoder`` overrides the header (``sd`` latents need the diffusion VAE;
    ``tiny`` thumbnails only Pillow).  A payload that is too short or too
    long - a cut-off or rambling prediction - is padded with zeros or
    truncated, and ``repaired`` says so.
    """
    header_encoder, width, height, payload, repaired = parse_text(text)
    name = (encoder or header_encoder).strip().lower()
    if name not in ("sd", "tiny"):
        raise VisionError(f"unknown image encoder {name!r} in the text; expected 'sd' or 'tiny'")
    if width % 8 or height % 8:
        raise VisionError(f"the image size {width}x{height} is not a multiple of 8")
    enc = get_encoder(name)
    shape = enc.latent_shape(width, height)
    length = shape[0] * shape[1] * shape[2]
    payload, changed = _fit_payload(payload, length)
    image = enc.decode(payload, width, height)
    return {
        "png": _png_bytes(image),
        "encoder": enc.name,
        "width": width,
        "height": height,
        "bytes": length,
        "repaired": repaired or changed,
    }

File: RadixCyclicNN/go/radixnet/vision.go
package radixnet

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"image"
	_ "image/gif"  // so load can read whatever the browser sends
	_ "image/jpeg" //
	"image/png"
	"regexp"
	"strconv"
	"strings"
)

// Images as text: the thumbnail encoder, the text format and the decoder.
//
// Stable Diffusion generates an image by decoding a latent - 4 x H/8 x W/8
// numbers - with its VAE, and vision.py runs that backwards.  The VAE needs
// torch and diffusers, so what is here is the other encoder: "tiny", an RGB
// thumbnail at 1/8 of the size (3 channels, one byte each), which is the same
// 8x spatial reduction and the same text format.
//
// One difference from Python is worth stating plainly, because it is the only
// place in this port where the two sides do not produce identical bytes: the
// *pixel reduction* differs.  Python's thumbnail goes through Pillow's Lanczos
// filter; this one uses a box filter, because reproducing Pillow's exact
// coefficients and rounding without Pillow to check against would be a
// guess.  So the same source image gives two different - equally valid - texts.
// Everything around that is shared: each side parses, trains on, predicts and
// decodes the other's texts, and an "sd" text can be parsed, measured and
// compared here too, it just cannot be decoded back into a picture without the
// weights.
//
//	img:tiny:128x128:AAECAwQF...

// Image encoder names; "auto" resolves to "tiny" here, because "sd" needs the
// diffusion weights and therefore the Python side.
var ImageEncoders = []string{"auto", "tiny"}

// DefaultImageSize is the square images are resized to before encoding.
const DefaultImageSize = 128

// MinImageSize and MaxImageSize bound it (and it must be a multiple of 8).
const (
	MinImageSize = 8
	MaxImageSize = 1024
)

const imageHeader = "img"

var imageTextRe = regexp.MustCompile(`(?s)^\s*img:([a-z0-9_]+):(\d+)x(\d+):(.*)$`)

// VisionError is an unreadable image, an unknown encoder or a text that is not
// an encoded image.
type VisionError struct{ Message string }

func (e *VisionError) Error() string { return e.Message }

func visionErrorf(format string, args ...any) error {
	return &VisionError{Message: fmt.Sprintf(format, args...)}
}

// PackImageText is img:<encoder>:<w>x<h>:<base64 of the quantised latent>.
func PackImageText(encoder string, width, height int, payload []byte) string {
	return fmt.Sprintf("%s:%s:%dx%d:%s", imageHeader, encoder, width, height,
		base64.StdEncoding.EncodeToString(payload))
}

// ParseImageText reads (encoder, width, height, payload, repaired) out of an
// encoded-image text.  The base64 is repaired when it is not clean - whitespace,
// stray characters, a missing tail, all of which a predicted text has - and the
// payload is not padded here (the decoder fits it to the latent it needs).
func ParseImageText(text string) (string, int, int, []byte, bool, error) {
	match := imageTextRe.FindStringSubmatch(text)
	if match == nil {
		return "", 0, 0, nil, false, visionErrorf("not an encoded image: expected 'img:<encoder>:<w>x<h>:<base64>'")
	}
	width, _ := strconv.Atoi(match[2])
	height, _ := strconv.Atoi(match[3])
	payload, repaired, err := RepairBase64(match[4])
	if err != nil {
		return "", 0, 0, nil, false, visionErrorf("%v", err)
	}
	if width <= 0 || height <= 0 {
		return "", 0, 0, nil, false, visionErrorf("invalid image size %dx%d", width, height)
	}
	return match[1], width, height, payload, repaired, nil
}

// checkImageSize validates a requested size.
func checkImageSize(size int) (int, error) {
	if size < MinImageSize || size > MaxImageSize || size%8 != 0 {
		return 0, visionErrorf("size must be a multiple of 8 between %d and %d, got %d",
			MinImageSize, MaxImageSize, size)
	}
	return size, nil
}

// NormaliseImageEncoder resolves "auto" and refuses what this side cannot do.
func NormaliseImageEncoder(name string) (string, error) {
	key := strings.ToLower(strings.TrimSpace(name))
	if key == "" || key == "auto" {
		return "tiny", nil
	}
	if key == "tiny" {
		return "tiny", nil
	}
	if key == "sd" {
		return "", visionErrorf("the Stable Diffusion encoder needs torch and diffusers: use the Python side " +
			"for it, or --encoder tiny here")
	}
	return "", visionErrorf("unknown image encoder %q; expected one of: %s", name, strings.Join(ImageEncoders, ", "))
}

// tinyLatentShape is (channels, H/8, W/8) of the thumbnail encoder.
func tinyLatentShape(width, height int) [3]int { return [3]int{3, height / 8, width / 8} }

// EncodedImage is what EncodeImage produces.
type EncodedImage struct {
	Text        string `json:"text"`
	Encoder     string `json:"encoder"`
	Width       int    `json:"width"`
	Height      int    `json:"height"`
	LatentShape []int  `json:"latent_shape"`
	Bytes       int    `json:"bytes"`
	Chars       int    `json:"chars"`
	SourceSize  []int  `json:"source_size"`
}

// EncodeImage turns image bytes (PNG, JPEG, GIF) into the text the network
// trains on: resized to size x size, reduced to a 1/8 RGB thumbnail, one byte
// per channel value, base64.
func EncodeImage(data []byte, size int, encoder string) (*EncodedImage, error) {
	if size == 0 {
		size = DefaultImageSize
	}
	size, err := checkImageSize(size)
	if err != nil {
		return nil, err
	}
	name, err := NormaliseImageEncoder(encoder)
	if err != nil {
		return nil, err
	}
	src, _, err := image.Decode(bytes.NewReader(data))
	if err != nil {
		return nil, visionErrorf("not a readable image: %v", err)
	}
	small := size / 8
	payload := resampleRGB(src, small, small)
	return &EncodedImage{
		Text: PackImageText(name, size, size, payload), Encoder: name, Width: size, Height: size,
		LatentShape: []int{3, small, small}, Bytes: len(payload), Chars: len(PackImageText(name, size, size, payload)),
		SourceSize: []int{src.Bounds().Dx(), src.Bounds().Dy()},
	}, nil
}

// resampleRGB box-filters src down to width x height and returns it as
// interleaved RGB bytes.  A box filter over the source pixels of each target
// cell is what a thumbnail is; it needs no dependencies and is stable, so the
// same image always gives the same text.
func resampleRGB(src image.Image, width, height int) []byte {
	bounds := src.Bounds()
	sw, sh := bounds.Dx(), bounds.Dy()
	out := make([]byte, 0, width*height*3)
	for y := 0; y < height; y++ {
		y0 := bounds.Min.Y + y*sh/height
		y1 := bounds.Min.Y + (y+1)*sh/height
		if y1 <= y0 {
			y1 = y0 + 1
		}
		for x := 0; x < width; x++ {
			x0 := bounds.Min.X + x*sw/width
			x1 := bounds.Min.X + (x+1)*sw/width
			if x1 <= x0 {
				x1 = x0 + 1
			}
			var rs, gs, bs, n uint64
			for py := y0; py < y1; py++ {
				for px := x0; px < x1; px++ {
					r, g, b, _ := src.At(px, py).RGBA() // 16-bit per channel
					rs += uint64(r >> 8)
					gs += uint64(g >> 8)
					bs += uint64(b >> 8)
					n++
				}
			}
			if n == 0 {
				n = 1
			}
			out = append(out, byte(rs/n), byte(gs/n), byte(bs/n))
		}
	}
	return out
}

// DecodedImage is what DecodeImageText produces.
type DecodedImage struct {
	PNG      []byte `json:"-"`
	Encoder  string `json:"encoder"`
	Width    int    `json:"width"`
	Height   int    `json:"height"`
	Bytes    int    `json:"bytes"`
	Repaired bool   `json:"repaired"`
}

// DecodeImageText turns an encoded - or predicted - text back into a PNG.  A
// payload that is too short or too long (a cut-off or rambling prediction) is
// padded with zeros or truncated, and Repaired says so.
func DecodeImageText(text, encoder string) (*DecodedImage, error) {
	headerEncoder, width, height, payload, repaired, err := ParseImageText(text)
	if err != nil {
		return nil, err
	}
	name := encoder
	if strings.TrimSpace(name) == "" {
		name = headerEncoder
	}
	if resolved, err := NormaliseImageEncoder(name); err != nil {
		return nil, err
	} else {
		name = resolved
	}
	if width%8 != 0 || height%8 != 0 {
		return nil, visionErrorf("the image size %dx%d is not a multiple of 8", width, height)
	}
	shape := tinyLatentShape(width, height)
	length := shape[0] * shape[1] * shape[2]
	payload, changed := fitPayload(payload, length)
	small := image.NewRGBA(image.Rect(0, 0, shape[2], shape[1]))
	for i := 0; i < shape[1]*shape[2]; i++ {
		small.Pix[i*4+0] = payload[i*3+0]
		small.Pix[i*4+1] = payload[i*3+1]
		small.Pix[i*4+2] = payload[i*3+2]
		small.Pix[i*4+3] = 255
	}
	full := image.NewRGBA(image.Rect(0, 0, width, height))
	for y := 0; y < height; y++ {
		sy := y * shape[1] / height
		for x := 0; x < width; x++ {
			sx := x * shape[2] / width
			at, to := (sy*shape[2]+sx)*4, (y*width+x)*4
			copy(full.Pix[to:to+4], small.Pix[at:at+4])
		}
	}
	var buf bytes.Buffer
	if err := png.Encode(&buf, full); err != nil {
		return nil, visionErrorf("cannot encode the PNG: %v", err)
	}
	return &DecodedImage{
		PNG: buf.Bytes(), Encoder: name, Width: width, Height: height, Bytes: length,
		Repaired: repaired || changed,
	}, nil
}

// fitPayload pads with zeros or truncates to length; returns (payload, changed).
func fitPayload(payload []byte, length int) ([]byte, bool) {
	if len(payload) == length {
		return payload, false
	}
	if len(payload) > length {
		return payload[:length], true
	}
	return append(append([]byte{}, payload...), make([]byte, length-len(payload))...), true
}

// DescribeVision reports what this side can do with images.
func DescribeVision() map[string]any {
	return map[string]any{
		"engine": "go", "encoders": ImageEncoders, "auto": "tiny", "default_size": DefaultImageSize,
		"sd": false, "resampling": "box",
		"resampling_note": "Python's thumbnail uses Pillow's Lanczos filter and this one a box filter, so the " +
			"same image gives a different (equally valid) text on each side; the format itself is shared, and " +
			"either side reads, trains on and decodes the other's texts",
		"sd_note": "the Stable Diffusion encoder needs torch and diffusers: use the Python server for it " +
			"(the text format is the same, so a model trained on sd texts by either side is read by both)",
		"text_format": imageHeader + ":<encoder>:<w>x<h>:<base64 of one byte per latent number>",
		"formats":     "PNG, JPEG and GIF",
	}
}
